### PR TITLE
[bug] fix empty read query df

### LINKED ISSUE #163: frames_from_hyper not pulling when there's no data in the table.
**Describe the bug**
I'm simply trying to pull all the frames from a hyper file (which I unfortunately can't give you).  But I think the problem is that one of the tables doesn't have any rows. Which then errors out [here](https://github.com/innobi/pantab/blob/f6a9fbb3f73cd51ada79e5d6d6aa114aea77dc98/pantab/_reader.py#L50) with 

``` py
ValueError: Length mismatch: Expected axis has 0 elements, new values have 60 elements
```

df in the line above is an empty dataframe.  

This is a problem because it's screwing up my other reads. 

**Expected behavior**
It gives me a blank dataframe with the proper columns but no data.  

Could skip it too, but that seems worse.  

**Desktop (please complete the following information):**
 - OS: windows10

### PR DESCRIPTION
**Pull request by abhinavDhulipala**
We've also ran into this problem. We believe pantab shouldn't error given an empty df.

As usual this is Red-Green tested... and covered

closes #163

### CODE REVIEW

**Comment by WillAyd** [comment]
Very nice! The OP mentioned that the issue stemmed from the C extension but this fix suggests that isn't the case. What line was actually throwing the error? It might be slightly preferable to just make the code organically handle the empty dataframe case instead of having to branch and return like this

**Comment by abhinavDhulipala** [comment]
So if you comment out the fix in readers and run `pytest pantab/tests/test_reader.py::test_empty_read_query`, you'll see something along the lines of...


```python
_____________________________________________________________________ test_empty_read_query _____________________________________________________________________

df =    int16  int32  int64  Int16  Int32  Int64  float32  float64  ...         int64_limits float32_limits float64_limits ....                    0            NaN            NaN       NaN    <NA>    <NA>     <NA>     <NA>

[3 rows x 23 columns]
tmp_hyper = PosixPath('/tmp/pytest-of-codespace/pytest-1/test_empty_read_query0/test.hyper')

    def test_empty_read_query(df, tmp_hyper):
        """
        red-green for empty query results
        """
        pantab.frame_to_hyper(df, tmp_hyper, table="test")
        query = "SELECT * FROM test limit 0"
>       result = pantab.frame_from_hyper_query(tmp_hyper, query)

pantab/tests/test_reader.py:123: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pantab/_reader.py:169: in frame_from_hyper_query
    return _read_query_result(result, None, use_float_na)
pantab/_reader.py:51: in _read_query_result
    df.columns = dtypes.keys()
venv/lib/python3.11/site-packages/pandas/core/generic.py:6218: in __setattr__
    return object.__setattr__(self, name, value)
properties.pyx:69: in pandas._libs.properties.AxisProperty.__set__
    ???
venv/lib/python3.11/site-packages/pandas/core/generic.py:767: in _set_axis
    self._mgr.set_axis(axis, labels)
venv/lib/python3.11/site-packages/pandas/core/internals/managers.py:227: in set_axis
    self._validate_set_axis(axis, new_labels)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = BlockManager
Items: RangeIndex(start=0, stop=0, step=1)
Axis 1: RangeIndex(start=0, stop=0, step=1), axis = 0
new_labels = Index(['int16', 'int32', 'int64', 'Int16', 'Int32', 'Int64', 'float32',
       'float64', 'bool', 'datetime64', 'datet...loat32_limits', 'float64_limits', 'non-ascii', 'boolean', 'string',
       'Float32', 'Float64'],
      dtype='object')

    @final
    def _validate_set_axis(self, axis: AxisInt, new_labels: Index) -> None:
        # Caller is responsible for ensuring we have an Index object.
        old_len = len(self.axes[axis])
        new_len = len(new_labels)
    
        if axis == 1 and len(self.items) == 0:
            # If we are setting the index on a DataFrame with no columns,
            #  it is OK to change the length.
            pass
    
        elif new_len != old_len:
>           raise ValueError(
                f"Length mismatch: Expected axis has {old_len} elements, new "
                f"values have {new_len} elements"
            )
E           ValueError: Length mismatch: Expected axis has 0 elements, new values have 23 elements

venv/lib/python3.11/site-packages/pandas/core/internals/base.py:85: ValueError
```

The actionable error here is `venv/lib/python3.11/site-packages/pandas/core/internals/base.py:85: ValueError`, which comes from line 50 (`df.columns = dtypes.keys()`) in `_reader.py`. This doesn't make sense since `len(df.columns) != len(dtypes)`. Another way we can handle this is by setting df to empty df with empty cols, i.e

```python
if df.empty:
    df = pd.DataFrame({col: [] for col in dtypes.keys()})
```

Personally, I prefer the approach I submitted, there is no need to execute the rest of the function and seems like an elegant enough solution. 

TL;DR I'm pretty sure the fix is purely a python problem, unless you'd rather the underlying C extension return df with empty columns. I think the decision you made of making the libpantab return an empty DF w/o initialized columns is the right choice, as it makes error handling like this organic. As usual, up to you. I'm fine either way. This seems less intrusive and cleaner IMHO

**Comment by WillAyd** [comment]
Thanks for all of the input. Overall with pantab we should aim to be consistent with how pandas `read_sql` works. It looks like that function returns an empty DataFrame of with all the columns of object-dtype when there is no data selected

**Comment by abhinavDhulipala** [comment]
I see, I'll put that in then

**Comment by abhinavDhulipala** [comment]
@WillAyd Addressed. It now tests that read_sql and and pantab read_query respond identically for the empty df case. This isn't asserted anywhere else though

**Comment by WillAyd** [comment]
Great work - thanks again for this! I'm loosely planning on cutting a new release next week to incorporate your recent changes. Let me know if that timing works or if there is other stuff you'd like to do before a new version

**Comment by abhinavDhulipala** [comment]
That timings perfect, appreciate it.
